core/src/saros/net/xmpp/Socks5ProxySupport.java

```diff
@@ -0,0 +1,363 @@
+    socks5Proxy.replaceLocalAddresses(Collections.emptyList());
+
+    log.info(
+        "started Socks5 proxy on port: "
+            + socks5Proxy.getPort()
+            + " [listening on all interfaces]");
+
+    final List<String> proxyAddressesToPublish = new ArrayList<>();
+
+    if (proxyAddresses != null && proxyAddresses.isEmpty())
+      log.warn("Socks5 preconfigured addresses list is empty, using autodetect mode");
+
+    if (proxyAddresses == null || proxyAddresses.isEmpty())
+      NetworkingUtils.getAllNonLoopbackLocalIPAddresses(true)
+          .stream()
+          .map(InetAddress::getHostAddress)
+          .forEach(proxyAddressesToPublish::add);
+    else proxyAddressesToPublish.addAll(proxyAddresses);
+
+    addSocks5ProxyAddresses(socks5Proxy, proxyAddressesToPublish, true);
+
+    if (stunService != null && stunServerAddress != null && !stunServerAddress.isEmpty()) {
+      try {
+        addressRetrieval.acquire(1);
+      } catch (InterruptedException e) {
+        log.warn("interupted while waiting to trigger a STUN discovery");
+        Thread.currentThread().interrupt();
+        return true;
+      }
+
+      ThreadUtils.runSafeAsync(
+          "Saros-stun-discovery",
+          log,
+          () -> {
+            try {
+              discoverAndPublishStunAddresses(newSocks5Proxy, stunServerAddress, stunServerPort);
+            } finally {
+              addressRetrieval.release(1);
+            }
+          });
+    }
+
+    if (upnpService != null && gatewayDeviceId != null && !gatewayDeviceId.isEmpty()) {
+      try {
+        addressRetrieval.acquire(1);
+      } catch (InterruptedException e) {
+        log.warn("interupted while waiting to trigger a UPNP port mapping");
+        Thread.currentThread().interrupt();
+        return true;
+      }
+
+      ThreadUtils.runSafeAsync(
+          "upnp-portmapping",
+          log,
+          () -> {
+            try {
+              mapPortAndPublishPublicAddress(
+                  newSocks5Proxy, gatewayDeviceId, useExternalGatewayDeviceAddress);
+            } finally {
+              addressRetrieval.release(1);
+            }
+          });
+    }
+
+    return true;
+  }
+
+  /**
+   * Stops the current Socks5 proxy if enabled and disables further usage of the Socks5 proxy. This
+   * method can be safely called to just prevent the global usage of the Socks5 proxy.
+   */
+  public synchronized void disableProxy() {
+
+    if (socks5Proxy == null) {
+      SmackConfiguration.setLocalSocks5ProxyEnabled(false);
+      return;
+    }
+
+    Socks5Proxy.getSocks5Proxy().stop();
+
+    SmackConfiguration.setLocalSocks5ProxyEnabled(false);
+
+    socks5Proxy = null;
+
+    if (socks5ProxyPort < 0) return;
+
+    log.info("stopped Socks5 proxy on port: " + socks5ProxyPort);
+
+    boolean wasInterrupted = false;
+
+    // we must wait here otherwise the device is not yet known
+
+    while (true) {
```

Comment:
I am not really sure about the behavior of "acquire". Would it lead to an eternal loop of "acquire" throws an InterruptedException, because the next call would also throw such an exception?